### DISCUSSION TITLE
Ability to rewrite URLs alongside next-intl rewrites to support A/B testing experimentation

### DISCUSSION THREAD

**Discussion by alexcarpenter** [amannn/next-intl#573] [Q&A]
### Is your feature request related to a problem? Please describe.

We are looking to introduce experimentation into a project already making use of next-intl. We're looking to use vercel edge config within middleware to rewrite the url based on which bucket the user falls into. We need some functionality within middleware to rewrite the URL and support the locale within the rewrite.

### Describe the solution you'd like

1. User visits `/experiment` route
2. User is in `test` bucket
3. Rewrite the url `/experiment/test`
4. This should also work on routes with the locale
5. `/en-GB/experiment` rewrites to `/en-GB/experiment/test`

### Describe alternatives you've considered

I've tried using `response.headers.set("x-middleware-rewrite", url.toString());` with no luck. This causes a 404.

**Comment by amannn** [comment]
Have you tried updating the pathname before passing the request to the `next-intl` middleware? Here's the official example: https://vercel.com/templates/next.js/cookies

I'll move this to a discussion since this seems to be a usage question.

**Reply by alexcarpenter** [reply]
That was what I was missing. Thank you so much!

**Comment by alexcarpenter** [comment]
@amannn is it possible to only update the pathname when its on a certain locale?

I am trying to update the pathname only on `en-US` routes, not `en-GB` route.

Basic example of what I am trying to do.

```js
const handleI18nRouting = createIntlMiddleware({
  locales,
  defaultLocale,
});

const bucket = 'test'

if (
  url.pathname === "/" &&
  bucket
) {
  request.nextUrl.pathname = `/${bucket}`;
}

const response = handleI18nRouting(request);

return response
```

This works for the `en-US` route but is also rewriting on `en-GB` route.

So I am seeing `/en-GB/test` when I need it to be just `/en-GB`.

I need to know the locale before the `handleI18nRouting` is called.

**Reply by alexcarpenter** [reply]
For reference this is the approach I am trying to use for AB testing, but only for `en-US` paths https://github.com/vercel/examples/blob/main/edge-middleware/ab-testing-statsig/middleware.ts

**Reply by amannn** [reply]
Sure, you can extract the `locale` from the incoming pathname and use it as part of a condition. I've just added [a rewrite example](https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/docs/routing/middleware#example-additional-rewrites) to the middleware docs of the release candidate, which are expected to be published soon.

Hope this helps!

**Reply by HarleySalas** [reply]
For me, doing this means that the default local will function as a rewrite whereas non-default locales function as a redirect;

ex:

User goes to auth guarded route: `/dashboard`, but does not have the necessary permissions.
Middleware following your instructions, rewrites to `/403`, but displaying `/dashboard` in the url bar. (expected outcome)

If the locale is anything other than the default locale, it functions as a redirect meaning for example:

User goes to auth guarded route: `/de/dashboard`, but is redirect to `/de/403`, instead of `/de/dashboard`, being rewritten to `/de/403` (undesired outcome)

I am of course using `localePrefix: "as-needed"` 

```
  if (pathname.startsWith("/dashboard") || pathname.startsWith("/admin")) {
    request.nextUrl.pathname = `${localePrefix}/403`;
  }
  ```
  
 `localePrefix` is either `/{locale}`, or an empty string, if there is no locale in the pathname (default locale).

I also tried adding the default locale prefix in every case as well, but it just results in behavior aligning with non-default locales. (functions as a redirect, instead of rewrite)
